### PR TITLE
Merge `BorrowKind::Unique` into `BorrowKind::Mut`

### LINKED ISSUE #112072: merge `BorrowKind::Unique` into `BorrowKind::Mut`
We currently have a separate `mir::BorrowKind` for the mutable borrows for closure upvars.

`BorrowKind::Unique` is an ordinary mutable borrow, requiring its target to be invariant, needing unique access, except that the target is not necessarily required to be marked as `mut` by the user.

We should merge this into `BorrowKind::Mut`, resulting in the following setup:
```rust
enum BorrowKind {
    Shared,
    Shallow,
    Mut { kind: MutBorrowKind },
}

enum MutBorrowKind {
    Default,
    TwoPhaseBorrow,
    ClosureCapture,
}
```
It is quite worrying to me that `fn BorrowKind::mutability` simply returns `Mutability::Not` for `BorrowKind::Unique`
https://github.com/rust-lang/rust/blob/99ff5afeb8717c096a4f153896e0c3856a3934ad/compiler/rustc_middle/src/mir/mod.rs#L2030-L2035

This already caused a bug in https://github.com/rust-lang/rust/issues/112056 where we used `PlaceContext::NonMutatingUse` for unique borrows, which resulted in us getting the wrong variance, cc #112070.

### PR DESCRIPTION
**Pull request by zirconium-n**
Fixes #112072 

Might have conflict with #112070

r? @lcnr 

I'm not sure what's the suitable change in a couple places.

### CODE REVIEW

File: compiler/rustc_const_eval/src/transform/check_consts/resolver.rs [hunk 1]
```diff
@@ -103,7 +103,7 @@ where
     fn ref_allows_mutation(&self, kind: mir::BorrowKind, place: mir::Place<'tcx>) -> bool {
         match kind {
             mir::BorrowKind::Mut { .. } => true,
-            mir::BorrowKind::Shared | mir::BorrowKind::Shallow | mir::BorrowKind::Unique => {
+            mir::BorrowKind::Shared | mir::BorrowKind::Shallow => {
```
Comment (zirconium-n) [inline compiler/rustc_const_eval/src/transform/check_consts/resolver.rs:106]: The logic changed here. Maybe I should not touch this?

File: compiler/rustc_const_eval/src/transform/check_consts/check.rs [hunk 2]
```diff
@@ -412,9 +412,6 @@ impl<'tcx> Visitor<'tcx> for Checker<'_, 'tcx> {
                         BorrowKind::Shallow => {
                             PlaceContext::NonMutatingUse(NonMutatingUseContext::ShallowBorrow)
                         }
-                        BorrowKind::Unique => {
```
Comment (zirconium-n) [inline compiler/rustc_const_eval/src/transform/check_consts/check.rs:415]: I think this is handled in #112070

File: compiler/rustc_middle/src/mir/mod.rs [hunk 3]
```diff
@@ -2029,22 +2029,24 @@ impl<'tcx> Rvalue<'tcx> {
 impl BorrowKind {
     pub fn mutability(&self) -> Mutability {
         match *self {
-            BorrowKind::Shared | BorrowKind::Shallow | BorrowKind::Unique => Mutability::Not,
+            BorrowKind::Shared | BorrowKind::Shallow => Mutability::Not,
             BorrowKind::Mut { .. } => Mutability::Mut,
```
Comment (zirconium-n) [inline compiler/rustc_middle/src/mir/mod.rs:2033]: Logic changed here.

File: compiler/rustc_middle/src/mir/mod.rs [hunk 4]
```diff
@@ -2029,22 +2029,24 @@ impl<'tcx> Rvalue<'tcx> {
...
         }
     }
 
     // FIXME: won't be used after diagnostic migration
     pub fn describe_mutability(&self) -> &str {
         match *self {
-            BorrowKind::Shared | BorrowKind::Shallow | BorrowKind::Unique => "immutable",
+            BorrowKind::Shared
+            | BorrowKind::Shallow
+            | BorrowKind::Mut { kind: MutBorrowKind::ClosureCapture } => "immutable",
```
Comment (zirconium-n) [inline compiler/rustc_middle/src/mir/mod.rs:2049]: I think we should return `"mutable"` when it's `MutBorrowKind::ClosureCapture`  here?

File: compiler/rustc_mir_build/src/check_unsafety.rs [hunk 5]
```diff
@@ -254,7 +254,9 @@ impl<'a, 'tcx> Visitor<'a, 'tcx> for UnsafetyVisitor<'a, 'tcx> {
                         );
                     };
                     match borrow_kind {
-                        BorrowKind::Shallow | BorrowKind::Shared | BorrowKind::Unique => {
```
Comment (zirconium-n) [inline compiler/rustc_mir_build/src/check_unsafety.rs:257]: I think `BorrowKind::Unique` actually belongs to `BorrowKind::Mut` ?

File: compiler/rustc_mir_build/src/check_unsafety.rs [hunk 6]
```diff
@@ -440,15 +442,19 @@ impl<'a, 'tcx> Visitor<'a, 'tcx> for UnsafetyVisitor<'a, 'tcx> {
                 visit::walk_expr(&mut visitor, expr);
                 if visitor.found {
                     match borrow_kind {
-                        BorrowKind::Shallow | BorrowKind::Shared | BorrowKind::Unique
```
Comment (zirconium-n) [inline compiler/rustc_mir_build/src/check_unsafety.rs:443]: Same to above.

**Review by zirconium-n — COMMENTED** [review]
I commented on places I'm not sure.

**Review by lcnr — COMMENTED** [review]
quickly skimmed it but need to take some time to think about the places where `Unique` currently differs from `Mut`, will only get to that next week probably

**Comment by zirconium-n** [comment]
> quickly skimmed it but need to take some time to think about the places where `Unique` currently differs from `Mut`, will only get to that next week probably

From my intuition, `Unique` is just a `Mut` except it does not require a mutable binding. (Maybe that's oversimplification?)

So here's some personal observation:

0. Diagnostics obviously should differ.

1. These ones should unify `Unique` and `Mut`. Since they are checking exactly whether it's a mutable borrow.
https://github.com/rust-lang/rust/blob/e29821ff85a2a3000d226f99f62f89464028d5d6/compiler/rustc_const_eval/src/transform/check_consts/resolver.rs#L103-L110
https://github.com/rust-lang/rust/blob/e29821ff85a2a3000d226f99f62f89464028d5d6/compiler/rustc_middle/src/mir/mod.rs#L2030-L2035
https://github.com/rust-lang/rust/blob/e29821ff85a2a3000d226f99f62f89464028d5d6/compiler/rustc_mir_build/src/check_unsafety.rs#L438-L454

2. Same to 1. But I think we can't create a `ref uniq` binding anyway so it probably does not matter?
https://github.com/rust-lang/rust/blob/e29821ff85a2a3000d226f99f62f89464028d5d6/compiler/rustc_mir_build/src/check_unsafety.rs#L247-L268

3. Same to 1. but this is just for diagnostic I guess.
https://github.com/rust-lang/rust/blob/e29821ff85a2a3000d226f99f62f89464028d5d6/compiler/rustc_middle/src/mir/mod.rs#L2044-L2050

4. This one probably stay different just to be safe.
https://github.com/rust-lang/rust/blob/e29821ff85a2a3000d226f99f62f89464028d5d6/compiler/rustc_const_eval/src/transform/promote_consts.rs#L456-L460

5. This one should stay different since this is the motivation to introduce `Unique`.
https://github.com/rust-lang/rust/blob/e29821ff85a2a3000d226f99f62f89464028d5d6/compiler/rustc_borrowck/src/lib.rs#L1958-L1969

File: compiler/rustc_borrowck/src/lib.rs [hunk 7]
```diff
@@ -1956,14 +1955,12 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
...
-            | Write(WriteKind::MutableBorrow(
-                borrow_kind @ (BorrowKind::Unique | BorrowKind::Mut { .. }),
-            )) => {
+            Reservation(WriteKind::MutableBorrow(borrow_kind @ BorrowKind::Mut { .. }))
+            | Write(WriteKind::MutableBorrow(borrow_kind @ BorrowKind::Mut { .. })) => {
                 let is_local_mutation_allowed = match borrow_kind {
-                    BorrowKind::Unique => LocalMutationIsAllowed::Yes,
+                    BorrowKind::Mut { kind: MutBorrowKind::ClosureCapture } => {
+                        LocalMutationIsAllowed::Yes
+                    }
```
Comment (lcnr) [inline compiler/rustc_borrowck/src/lib.rs:1963]: can you add a comment here that this is used for "mutable variable used immutably" which is the only behavior difference between `ClosureCapture`s and mutable borrows?

File: compiler/rustc_const_eval/src/transform/promote_consts.rs [hunk 8]
```diff
@@ -457,7 +457,9 @@ impl<'tcx> Validator<'_, 'tcx> {
         match kind {
             // Reject these borrow types just to be safe.
             // FIXME(RalfJung): could we allow them? Should we? No point in it until we have a usecase.
-            BorrowKind::Shallow | BorrowKind::Unique => return Err(Unpromotable),
+            BorrowKind::Shallow | BorrowKind::Mut { kind: MutBorrowKind::ClosureCapture } => {
+                return Err(Unpromotable);
+            }
```
Comment (lcnr) [inline compiler/rustc_const_eval/src/transform/promote_consts.rs:462]: please exhaustively match on the remaining two kinds in `BorrowKind::Mut` and add a FIXME for a future PR:

> consider changing this to only promote `&mut []` for default borrows, also forbidding two phase borrows

File: compiler/rustc_middle/src/mir/mod.rs (hunk 4 above)
Comment (lcnr) [reply compiler/rustc_middle/src/mir/mod.rs:2049]: this function is completely unused. You can remove it entirely iirc

File: compiler/rustc_mir_build/src/check_unsafety.rs (hunk 5 above)
Comment (lcnr) [reply compiler/rustc_mir_build/src/check_unsafety.rs:257]: yes, it should always require `unsafe`, the status quo is wrong

**Review by lcnr — COMMENTED** [review]
I am not sure whether specialcasing `ClosureCapture` borrows this much in compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs is necessary or even particularily helpful, but :shrug: 

after dealing with the comments, r=me 

thank you for working on this :heart:

File: compiler/rustc_borrowck/src/diagnostics/mod.rs [hunk 9]
```diff
@@ -623,9 +623,9 @@ impl UseSpans<'_> {
                     Some(kd) => match kd {
                         rustc_middle::mir::BorrowKind::Shared
                         | rustc_middle::mir::BorrowKind::Shallow
-                        | rustc_middle::mir::BorrowKind::Unique => {
-                            CaptureVarKind::Immut { kind_span: capture_kind_span }
-                        }
+                        | rustc_middle::mir::BorrowKind::Mut {
+                            kind: rustc_middle::mir::MutBorrowKind::ClosureCapture,
+                        } => CaptureVarKind::Immut { kind_span: capture_kind_span },
```
[the remark this answers is not in the document]
Comment (zirconium-n) [reply compiler/rustc_borrowck/src/diagnostics/mod.rs]: Yes,  I agree we should change this to `CaptureVarKind::Mut`.  I missed this place.

As for what happens, nothing observable happens.

**Comment by zirconium-n** [comment]
@rustbot label -S-waiting-on-author +S-waiting-on-review

> we should also run crater after the remaining changes as I expect the unsafeck and const checks to be theoretically breaking. Don't have the capacity to try and get an actual test for that though

Const checks should be ok, since we forbid `&mut` in const so it's rejected anyway?
Not sure about unsafeck.

File: compiler/rustc_middle/src/mir/syntax.rs [hunk 10]
```diff
@@ -182,6 +182,16 @@ pub enum BorrowKind {
...
 
+    /// Data is mutable and not aliasable.
+    Mut { kind: MutBorrowKind },
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, TyEncodable, TyDecodable)]
+#[derive(Hash, HashStable)]
+pub enum MutBorrowKind {
+    Default,
+    /// this borrow arose from method-call auto-ref. (i.e., `adjustment::Adjust::Borrow`)
```
Comment (lcnr) [inline compiler/rustc_middle/src/mir/syntax.rs:193]:
```suggestion
    /// This borrow arose from method-call auto-ref. (i.e., `adjustment::Adjust::Borrow`)
```

**Comment by lcnr** [comment]
@craterbot check

**Comment by lcnr** [comment]
looked at a bunch and they all looked spurious, didn't get a nice testcase for them

r=me after rebase

**Pull request merged**